Add maximum contrast function

lexun opened the pull request:
Hello and thanks for the color extras! I've been using this library in a project an have found it very useful.

When generating background colors dynamically, I discovered the need to know if I should use a light or dark text color to best contrast the background. I've implemented a function using the w3c contrast-ratio and relative luminance formulas, and thought it might be useful in this library somewhere.

This PR includes the code I've been using, but I'm happy to change the API or naming in any way that makes it more useable. Just let me know if you have any ideas you'd like me to try.

Here is a visual of the results currently:

![image](https://cloud.githubusercontent.com/assets/1910143/22721268/b3c62bc2-ed53-11e6-9535-819e8731ffdd.png)



##### For reference:
- https://www.w3.org/TR/WCAG20/#contrast-ratiodef
- https://www.w3.org/TR/WCAG20/#relativeluminancedef

lexun committed:
Add color to contrast function

eskimoblood commented:
Thanks for adding feature, I will take a look at the code later today.

lexun commented:
Great, thank you. It occurs to me that it might be better to split this into a few lower level functions, one that returns luminance as a float, and another that determines contrast ratio.

What I have now was designed specifically to determine if I should use one of my light or dark text colors on a given background. Maybe just providing functions for luminance and contrast ratio separately would keep it a little lower level and serve more use cases?

If so, does it still make sense to build on the luminance and contrast ratio functions with something that would answer the question of light or dark text? Instead of just returning `Light` or `Dark` we could even have a function that takes a base color and a list of colors, returning the one from the list with the highest contrast to the base color.

Let me know what you think.

eskimoblood commented:
I think splitting it into more lower level would make sense, feel free to update this pull request.

lexun committed:
Add luminance and contrast ratio functions

Replaces colorToContrast with lower level functions to provide more
flexibility in serving other use cases.

lexun commented:
Updated the PR to only expose functions for calculating luminance and contrast ratio. The downside is that now there isn't a function that directly solves the problem of which text color to use on a given background. Using the `contrastRatio` function would make that pretty trivial though.

I was thinking that a highest contrast function would be useful, something like:
```elm
highestContrast : Color -> List Color -> Color
```

So, of your list of possible text colors it could return the one with the highest contrast to the first color you provide. I think that is a pretty flexible way to solve the initial problem:

```elm
idealTextColor = highestContrast backgroundColor [Color.black, Color.white]
```

If we do want that, I'm not sure what module it would belong in. Thoughts?

eskimoblood commented on src/Color/Convert.elm:
```diff
@@ -303,3 +316,62 @@ xyzToColor { x, y, z } =
+luminance : Color -> Float
+luminance cl =
+    let
+        { r, g, b } =
```
As small improvement, this can be writen as:

```
(r,g,b) = (toRgb >> (\(c) -> (f c.red,f c.green,f c.blue))) cl
```

eskimoblood reviewed (commented):
If you want to add this feature I would suggest to have an Accessibility package, that could also have methods to simulate colour blindness

lexun committed:
Improve rgb assignment function in luminance

lexun commented:
Sounds good, I'll see about starting on the Accessibility module with the contrast function tomorrow.

lexun committed:
Add Accessibility module

Move in luminance and contrast ratio

lexun committed:
Add maximum contrast function

lexun committed:
Add module docs for accessibility

eskimoblood commented on src/Color/Accessibility.elm:
```diff
@@ -0,0 +1,83 @@
+maximumContrast base options =
+    let
+        compareContrast c1 c2 =
+            flip compare
```
Why you need to use `flip` here, isn't it the same as
```
compare (contrastRatio base c2) (contrastRatio base c1)
```

lexun replied:
Yeah, I just wanted to draw attention to the fact that the list was going to be in reverse order. I'm happy to change it if you think that's unnecessary.

lexun replied:
I suppose it draws more question than clarity, will fix.

eskimoblood reviewed (commented):
Looks good, thanks again for your help, please have a look add the comment about `flip` after that I will merge it and release a new version.

lexun committed:
Remove unnecessary flip

lexun committed:
Add Accessibility to exposed modules

lexun committed:
Add Accessibility to readme

eskimoblood commented:
The new version is published on elm packages, thanks again for the pull request

lexun commented:
Great, thanks! Really appreciate all the feedback along the way.
